app/javascript/components/displays.jsx

```diff
@@ -0,0 +1,214 @@
+import Weather from '@widgets/weather';
+import Traffic from '@widgets/traffic';
+
+const Wrapper = styled.div`
+  outline: none;
+`;
+
+const widgetElements = {
+  Weather,
+  Twitter,
+  Numbers,
+  Traffic,
+};
+
+const prefetchDelay = 1500;
+
+export const WidgetDisplay = ({
+  stateCallback,
+  superState,
+  cityName,
+  data,
+}) => {
+  const lastPrefetchTime = useRef(0);
+  const timerId = useRef(null);
+
+  const { location } = data;
+  const { widgets } = location;
+  const { enabled: enabledWidgets, byIdOrFirst: current } = widgets;
+
+  const WidgetElement = widgetElements[current.name].Panel;
+
+  const index = enabledWidgets.findIndex(w => w.id === current.id);
+
+  const prevWidgetId =
+    enabledWidgets[mathMod(index - 1, enabledWidgets.length)].id;
+
+  const nextWidgetId =
+    enabledWidgets[mathMod(index + 1, enabledWidgets.length)].id;
+
+  const setNextWidgetId = () => {
+    stateCallback(({ transitionTime, pausedTime }) => ({
+      currentWidgetId: nextWidgetId,
+      transitionTime,
+      pausedTime,
+    }));
+  };
+
+  const setPrevWidgetId = () => {
+    stateCallback(({ transitionTime, pausedTime }) => ({
+      currentWidgetId: prevWidgetId,
+      transitionTime,
+      pausedTime,
+    }));
+  };
+
+  const switchPages = ({ keyCode }) => {
+    const upKey = 38;
+    const downKey = 40;
+    const SPACEBAR = 32;
+    const { pausedTime } = superState;
+    if (keyCode === SPACEBAR) {
+      if (pausedTime) {
+        stateCallback(({ currentWidgetId, transitionTime }) => ({
+          currentWidgetId,
+          pausedTime: null,
+          transitionTime: transitionTime - pausedTime + Date.now(),
+        }));
+      } else {
+        stateCallback(({ currentWidgetId, transitionTime }) => ({
+          currentWidgetId,
+          transitionTime,
+          pausedTime: Date.now(),
+        }));
+      }
+    }
+    if (keyCode === downKey) {
+      setNextWidgetId();
+    }
+    if (keyCode === upKey) {
+      setPrevWidgetId();
+    }
+  };
+
+  const handleClicks = ({ target }) => {
+    const targetId = target.id;
+    if (targetId.startsWith('widget_')) {
+      const widgetId = parseInt(targetId.replace('widget_', ''), 10);
+      stateCallback({
+        currentWidgetId: widgetId,
+      });
+    }
+  };
+
+  useEffect(() => {
+    timerId.current = setInterval(() => {
+      const { transitionTime, pausedTime } = superState;
+      if (!transitionTime || pausedTime) {
+        return;
+      }
+      const prefetchTime = transitionTime - prefetchDelay;
+      const now = Date.now();
+
+      if (transitionTime < now) {
+        setNextWidgetId();
+      } else if (
+        prefetchTime < now &&
+        lastPrefetchTime.current < now - prefetchDelay
+      ) {
+        lastPrefetchTime.current = now;
+        setNextWidgetId();
+      }
+    }, 500);
+
+    return () => {
+      if (timerId.current) {
+        clearInterval(timerId.current);
+      }
+    };
+  }, [superState]);
```

Comment:
The WidgetDisplay has information about enabledWidgets from data that is necessary for setting the next Widget. This is why the componentDidMount was moved to the WidgetDisplay.

Regarding your question, without the dependency superState the WidgetDisplay could be paused (by spacebar)  but the timer would still be running (so the widgets would still switch even if paused).

The interval for timerId is set up with the superState which is _passed_ from WidgetController - the superState remains the same in WidgetDisplay even though WidgetController state may have changed.  I suspect this means the superState is outdated in setInterval, and thus the interval needs to be reset with a new superState on every re render of WidgetController.

Is this ok or should it be changed?

Reply:
Ok, that makes more sense but it seems like we introduced a new edge case. Our setInterval is destroyed and created on updates to state, not just component mount. This could mean destroying in interval that has partially run (say 250ms from start) and shift all subsequent checks. We may need to rethink approach or increase the frequency of our checks to account for the new edge case.